### PR TITLE
Shortest path directions

### PR DESCRIPTION
**Pull request by ethorondil**
Adds a new command, _dirs, that prints the shortest path to rooms satisfying a condition.

By default it searches the room name, but different flags can give different options.  The format is
`_dirs [-(name|desc|note|exits|all)] search pattern`.

For example,
```
>_dirs caras
Distance 19.45: The Gates of Caras Galadhon
dirs: e3n4e2ne
Distance 20.05: Before the Walls of Caras Galadhon
dirs: e3n2ene3n2e
>_dirs -a ferry
Distance 14.9: Landing in the Forest
dirs: 2en3ese
Distance 64.75: A Landing
dirs: 2s6e
Distance 586.45: Road to Standelf
dirs: nwn6ws3wn6w2s4wsw2sw3s8ws9wn4w11nwn2wu8wsw3de3dn8ws8wswsw2usds3wn3wnws3wu6wswnwnwnwnwn6ws5w2sw4s3wsws2ws3wsws4wsws3ws3wsws4ws2wswsws10ws2wn6wn2wn6wsws4wnwnwnwn3wnw2n9ws3w4s4w2s2ws2ws6wnwnwnwnwnw2n2e2n2w2ne10nwnwnwnwnwn2wnwnw2nwnwnwnwn3wnwn3wn7wn2wn2w2nwn4w3n2w2nw5nw2nw8nw3ne3ne2nene2ne4ne5nenene2ne4ne4ne4n2wnw6nw5n2w4n3wn2wn6wn5ws4wswsw4sws2ws
[...]
```
(Why does the path to the ferry jog north? Because doing so avoids a noride room.)

or you can search for a specific exit, e.g.:
```
>_dirs -e hatch
Distance 13.75: An Abandoned Cellar
dirs: wdse8sw2usw
Distance 16.25: Old Ruins
dirs: wdse8sw2uswu
[...]
```
or search your notes with `_dirs -n` or `_dirs -note`.

The last commit here adds a delay to offline character moving, giving it time to sync up when speedwalking.

[This is a pretty big commit, so you probably want me to reorganize it/conform to the style more; just let me know how.  I also have a similar _search command to highlight rooms, which shares the filter code here and for which I'll submit a PR once this lands.]

### CODE REVIEW

File: src/parser/abstractparser.cpp [hunk 1]
```diff
@@ -856,6 +925,21 @@ bool AbstractParser::parseUserCommands(QString& command)
     else
       if (str.startsWith("_block"))   {dooraction = true; daction = DAT_BLOCK;}
 
+    if (str.startsWith("_dirs"))
+    {
+      QString pattern_str = str.section(' ', 1).trimmed();
+      if(pattern_str.size() == 0)
+      {
+        emit sendToUser("Usage: _search <pattern in quotes> [section to search]\r\n");
```
Comment (nschimme) [inline src/parser/abstractparser.cpp]: Should be _dirs rather than _search ?

File: src/parser/abstractparser.cpp [hunk 2]
```diff
@@ -856,6 +925,21 @@ bool AbstractParser::parseUserCommands(QString& command)
     else
       if (str.startsWith("_block"))   {dooraction = true; daction = DAT_BLOCK;}
 
+    if (str.startsWith("_dirs"))
```
Comment (nschimme) [inline src/parser/abstractparser.cpp]: The _dirs command and usage should also be added to the _help prompt

File: src/parser/abstractparser.cpp [hunk 3]
```diff
@@ -1201,6 +1285,7 @@ void AbstractParser::offlineCharacterMove(CommandIdType direction)
     }
   }
   m_mapData->unselect(rs1);
+  usleep(40000);
```
Comment (nschimme) [inline src/parser/abstractparser.cpp]: Any reason why we needed this usleep?

File: src/mapdata/roomfilter.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#include "roomfilter.h"
+#include <assert.h>
+#include <errno.h>
+
+const char * RoomFilter::parse_help = "Parse error; format is: [-(name|desc|note|exits|all)] text\r\n";
```
Comment (nschimme) [inline src/mapdata/roomfilter.cpp]: This should probably also be documented somehow. Or maybe just mention _dirs  -help ?

File: src/parser/abstractparser.cpp (hunk 3 above)
Comment (ethorondil) [inline src/parser/abstractparser.cpp]: This is a hackish fix to the following issue: in trunk, speedwalking doesn't work offline.  This makes it work (at least for me, at my computer's speed).

I don't really understand the root of the issue.  This makes it a little bit slow for long journeys, but at the same time it lets you watch the path taken, so it's not too bad.

**Comment by ethorondil** [comment]
OK, I've fixed the easy documentation issues.  Since it's hard to explain everything in one line, it might be good to have a place for more extensive documentation, like `_help _dirs` or something.

File: src/parser/abstractparser.cpp (hunk 3 above)
Comment (nschimme) [inline src/parser/abstractparser.cpp]: Seems fine to me since it's only limited to offline character moves.

**Comment by nschimme** [comment]
:+1:

**Pull request merged**